contracts/zkTrueUp/extensions/pendlePT/PTWithRedStonePriceFeed.sol

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+import {PendlePYLpOracle} from "@pendle/core-v2/contracts/oracles/PendlePYLpOracle.sol";
+import {PendlePYOracleLib} from "@pendle/core-v2/contracts/oracles/PendlePYOracleLib.sol";
+import {IPMarket} from "@pendle/core-v2/contracts/interfaces/IPMarket.sol";
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+
+/**
+ * @title The customized Pendle PT price feed contract mutated from Chainlink AggregatorV3Interface
+ * @author Term Structure Labs
+ * @notice Use the customized price feed contract to normalized price feed interface for Term Structure Protocol
+ */
+contract PTWithRedStonePriceFeed is AggregatorV3Interface {
+    using Math for uint256;
+    using SafeCast for *;
+    using PendlePYOracleLib for IPMarket;
+
+    // Pendle PY LP oracle, refer to `https://docs.pendle.finance/Developers/Oracles/HowToIntegratePtAndLpOracle`
+    PendlePYLpOracle public immutable PY_LP_ORACLE;
+    // Pendle market
+    IPMarket public immutable MARKET;
+    // TWAP duration
+    uint32 public immutable DURATION;
+    // PT to asset rate base (i.e. 1e18 for decimals 18)
+    uint256 internal immutable PT_TO_ASSET_RATE_BASE;
+    // RedStone price feed interface
+    AggregatorV3Interface public immutable REDSTONE_PRICE_FEED;
+
+    // error to call `getRoundData` function
+    error GetRoundDataNotSupported();
+    // error when Pendle PY LP oracle is not ready
+    error OracleIsNotReady();
+
+    /**
+     * @notice Construct the PT price feed contract
+     * @param pendlePYLpOracle The Pendle PY LP oracle contract
+     * @param market The Pendle market contract
+     * @param duration The TWAP duration
+     * @param ptToAssetRateBase The PT to asset rate base
+     * @param redStonePriceFeed The RedStone price feed interface
+     */
+    constructor(
+        PendlePYLpOracle pendlePYLpOracle,
+        IPMarket market,
+        uint32 duration,
+        uint256 ptToAssetRateBase,
+        AggregatorV3Interface redStonePriceFeed
+    ) {
+        PY_LP_ORACLE = pendlePYLpOracle;
+        MARKET = market;
+        DURATION = duration;
+        PT_TO_ASSET_RATE_BASE = ptToAssetRateBase;
+        REDSTONE_PRICE_FEED = redStonePriceFeed;
```

Comment:
Is there a way to validate SY price feed?

Reply:
I added a check to ensure the price from redstone should be non-zero, but for the validity of the SY price, it seems that we can only trust it to be correct if `oracleIsReady`.